Remove optim in favor of independent golden section implementation

piever opened the pull request:
Fix #83. This adds a simple implementation of the [golden section algorithm](https://en.wikipedia.org/wiki/Golden-section_search), so that KernelDensity can drop the Optim dependency. The issue (at least for me) is that Optim is somewhat heavy (3 seconds of load time), and KernelDensity is needed as a dependency for plotting packages for 1D and 2D density plots, so ideally it should be a bit more lightweight.

Removing Optim shaves off a couple of seconds from load time.

On master:

```julia
julia> @time using KernelDensity
  6.803513 seconds (12.15 M allocations: 706.212 MiB, 2.90% gc time)
```

With this PR:

```julia
julia> @time using KernelDensity
  4.620525 seconds (7.87 M allocations: 475.978 MiB, 0.72% gc time)
```

piever commented:
Bump?

tpapp commented:
This looks ok to me at a first glance but I haven't had time for a detailed review, sorry. Will try to do it soon.

piever commented:
Sorry to bump this again, but it seems that it is important to get this in to move forward with changes in Makie. We are adding support to statistical plots in AbstractPlotting.jl (see https://github.com/JuliaPlots/AbstractPlotting.jl/pull/478), and it looks like this PR mitigates the increase in load time substantially.

Here are times to load (after a first run to precompile) on julia 1.5

```julia
# AbstractPlotting master 
julia> @time using AbstractPlotting
  6.322670 seconds (9.76 M allocations: 600.265 MiB, 1.30% gc time)

# AbstractPlotting with pv/stats PR and KernelDensity release
julia> @time using AbstractPlotting
 13.004530 seconds (19.41 M allocations: 1.097 GiB, 3.11% gc time)

# AbstractPlotting with pv/stats PR and this PR on KernelDensity
julia> @time using AbstractPlotting
 10.147386 seconds (15.28 M allocations: 905.327 MiB, 2.54% gc time)
```

(Removing Optim seems like the only easy way to speed things up, because it is not used by most other statistical packages, whereas most other deps are shared and would be much harder to remove.)

tpapp commented on src/univariate.jl:
```diff
@@ -173,6 +173,52 @@ function kde(data::RealVector; bandwidth=default_bandwidth(data), kernel=Normal,
     kde(data,dist;boundary=boundary,npoints=npoints,weights=weights)
 end
 
+function optimize(f, x_lower, x_upper; iterations=1000, rel_tol=nothing, abs_tol=nothing)
```
Please write a brief docstring, explaining what the function does, the parameters, and what is returned; mentioning that it is not exported.

tpapp commented on src/univariate.jl:
```diff
@@ -173,6 +173,52 @@ function kde(data::RealVector; bandwidth=default_bandwidth(data), kernel=Normal,
+    rtol = something(rel_tol, sqrt(eps(T)))
+    atol = something(abs_tol, eps(T))
+    
+    function acceptable_interval(lower, upper)
```
```suggestion
    function midpoint_and_convergence(lower, upper)
```

tpapp commented on src/univariate.jl:
```diff
@@ -173,6 +173,52 @@ function kde(data::RealVector; bandwidth=default_bandwidth(data), kernel=Normal,
+    function acceptable_interval(lower, upper)
+        midpoint = (lower + upper) / 2
+        tol = atol + rtol * midpoint
+        return (upper - lower) <= 2tol
```
```suggestion
        midpoint, (upper - lower) <= 2tol
```

tpapp commented on src/univariate.jl:
```diff
@@ -173,6 +173,52 @@ function kde(data::RealVector; bandwidth=default_bandwidth(data), kernel=Normal,
+    for _ in 1:1000
+        h *= invphi
+        if fc < fd
+            acceptable_interval(a, d) && break
```
```suggestion
            m, converged = midpoint_and_convergence(a, d)
            converged && return m
```

tpapp commented on src/univariate.jl:
```diff
@@ -173,6 +173,52 @@ function kde(data::RealVector; bandwidth=default_bandwidth(data), kernel=Normal,
+            c = a + invphisq * h
+            fc = f(c)
+        else
+            acceptable_interval(c, b) && break
```
```suggestion
            m, converged = midpoint_and_convergence(c, b)
            converged && return m
```

tpapp commented on src/univariate.jl:
```diff
@@ -173,6 +173,52 @@ function kde(data::RealVector; bandwidth=default_bandwidth(data), kernel=Normal,
+        end
+    end
+
+    return fc < fd ? (a + d) / 2 : (c + b) / 2
```
```suggestion
    error("Reached maximum number of iterations without convergence.")
```

tpapp commented on test/univariate.jl:
```diff
@@ -63,3 +63,9 @@ end
+rel_tol = sqrt(eps(Float64))
+abs_tol = eps(Float64)
+
+minimizer = KernelDensity.optimize(x -> (x - 1)^2, 0, 2)
```
```suggestion
minimizer = @inferred KernelDensity.optimize(x -> (x - 1)^2, 0, 2)
```

tpapp commented:
Sorry for not reviewing earlier, I appreciate the ping. I requested some minor changes, and will merge ASAP when this converges.

piever commented:
Thank you for the review! I've addressed the feedback.

> I requested some minor changes, and will merge ASAP when this converges.

Great! Could it be possible to also have a release after this is merged?
